### PR TITLE
Avoid cases in OOP of accessing .Solution.Workspace

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
We intend to ban this on the OOP side so that services are pure data and so we're not:

1. exposing a mutable concept which has no actually supported ways to mutate it.
2. having to create and maintain a complex object with complex syncing code.

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ProjectExtensions.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ProjectExtensions.cs:20]: code style does some funky stuff that we don't need in OOP.  temporarily splitting this behavior so we don't go throught eh workspace here.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Workspace/Mef/MefWorkspaceServices.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Workspace/Mef/MefWorkspaceServices.cs:63]: can't go to this level yet.  far too many workspace services immediately grab the workspace on construction.  will need to work through avoiding that.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Workspace/Mef/MefWorkspaceServices.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Workspace/Mef/MefWorkspaceServices.cs:63]: woudl like to enable this, but it fires too many times currently.  WIP on this part :)

File: src/Workspaces/Core/Portable/Shared/Extensions/ProjectExtensions.cs
Comment (tmat) [inline src/Workspaces/Core/Portable/Shared/Extensions/ProjectExtensions.cs:16]: >IsFromPrimaryBranch [](http://example.com/codeflow?start=27&length=19)

The extension method doesn't seem that useful now.

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs
Comment (tmat) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs:50]: >PartialSemanticsEnabled [](http://example.com/codeflow?start=22&length=23)

What is partial semantics?

**Comment by tmat** [comment]
>             => project.Solution.Workspace.Services.GetExtendedLanguageServices(project.Language);

Can't we use SolutionServices here?

---
Refers to: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ProjectExtensions.cs:37 in 9efa885. [](commit_id = 9efa8850af1d8db0e95667761d06ba5be76ad26c, deletion_comment = False)

File: src/Workspaces/Core/Portable/Shared/Extensions/ProjectExtensions.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Extensions/ProjectExtensions.cs:16]: trying to not break any semantics here :)

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs:50]: partial semantics are a way for features to say: give me a snapshot of whatever data you have right now, even if it's incomplete.  e.g. if you still have 1000 files to load/parse, i'd prefer to just pretend those files don't exist, rather than waiting on them.  It's for features that have extremely tight latency requirements (E.g. completion).

Generally speaking, this means during the start of VS you might observe some incomplete results (though it's rare to notice) but it means you don't have long delays on completion/commit.

Eventually, the workspace has caught up and you will then get the benefits of speed and completeness.  But this functionality allows features that care more about speed to let the workspace know.

**Comment by CyrusNajmabadi** [comment]
> Can't we use SolutionServices here?

I can't find what line of code this is reffering to to answer :)

File: src/Workspaces/Core/Portable/Shared/Extensions/ProjectExtensions.cs
Comment (tmat) [reply src/Workspaces/Core/Portable/Shared/Extensions/ProjectExtensions.cs:16]: Inlining this wouldn't break any.

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs:50]: Why would we disable this per workspace?

File: src/Workspaces/Core/Portable/Shared/Extensions/ProjectExtensions.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Shared/Extensions/ProjectExtensions.cs:16]: oh.  i see what you mean.  that's just not really something i'm concerned about right now :)  i'm just trying first and foremost to decouple code from jumping from solution-to-workspace, particularly in the remote case.

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs:50]: The general appraoch in the VS workspace is tehre is a 'bg compiler' that is cosntantly pulling on the data and bringing it up to completeness.  So if something asks for partial semantics, they know they're getting closer and closer to a 'complete' view, jsut that it will always be fast to request it.

in a workspace that doesn't have such a component running, then these components may continually see a *highly incomplete* view that never converges on completeness.  So this is an opt in systems for hosts that support this concept.  MSBuildWorkspace, for example, does not.

**Comment by tmat** [comment]
ProjectExtensions.GetExtendedLanguageServices

**Comment by CyrusNajmabadi** [comment]
Unfortunately no.  It returns HostLanguageServices which is the existing, mutable-exposing API.   We could have like an GetExtended*Project*Services.  but it wouldn't do anything currently since only CodeStyle has different behavior here.  And COdeStyle can't use Solution/ProjectServices yet until we get API approval to make those public (which unblocks a lot more cases for me to go through).

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs:50]: I see. Would you mind adding these comments to the code?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs:50]: sure.  can i add in followup?

**Review by tmat — APPROVED** [review]

**Pull request merged**
